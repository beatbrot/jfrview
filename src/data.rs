use std::any::type_name;
use anyhow::{anyhow, Context};
use jfrs::reader::{
    event::{Accessor, Event},
    value_descriptor::ValueDescriptor,
    JfrReader,
};
use log::debug;
use std::io::{Read, Seek};
use jfrs::reader::value_descriptor::Primitive;

pub const EXEC_SAMPLE: &str = "jdk.ExecutionSample";

pub const NATIVE_EXEC_SAMPLE: &str = "jdk.NativeMethodSample";

#[derive(Debug)]
#[allow(dead_code)]
pub struct ExecutionSample {
    pub start_time: i64,
    pub thread: Thread,
    pub stack_trace: StackTrace,
    pub native: bool,
}

impl ExecutionSample {
    pub fn visit_events<T: Read + Seek>(
        source: T,
        mut visitor: impl FnMut(ExecutionSample),
    ) -> anyhow::Result<()> {
        let mut reader = JfrReader::new(source);

        for result in reader.chunks() {
            let (mut r, c) = result.with_context(|| "Unable to parse chunk.")?;
            debug!("Parsing chunk with size {}", c.header.chunk_size);
            for event_res in r.events(&c) {
                let event = event_res.with_context(|| "Unable to parse event.")?;
                if event.class.name() == EXEC_SAMPLE || event.class.name() == NATIVE_EXEC_SAMPLE {
                    let sample = ExecutionSample::from(event);
                    if !sample.stack_trace.truncated {
                        visitor(sample);
                    }
                }
            }
        }

        Ok(())
    }
}

impl From<Event<'_>> for ExecutionSample {
    fn from(value: Event) -> Self {
        let native = value.class.name() == NATIVE_EXEC_SAMPLE;
        let v = value.value();
        let start_time: i64 = extract_primitive(&v, "startTime");
        let thread: Thread = value.value().get_field("sampledThread").unwrap().into();
        let stack_trace: StackTrace = value.value().get_field("stackTrace").unwrap().into();
        Self {
            start_time,
            thread,
            stack_trace,
            native,
        }
    }
}

#[derive(Debug)]
#[allow(dead_code)]
pub struct Thread {
    pub java_name: Option<String>,
    pub java_thread_id: i64,
}

impl From<Accessor<'_>> for Thread {
    fn from(value: Accessor<'_>) -> Self {
        let java_name = extract_nullable_str(&value,"javaName");
        let java_thread_id: i64 = extract_primitive(&value, "javaThreadId");
        Thread {
            java_name,
            java_thread_id,
        }
    }
}

impl PartialEq for Thread {
    fn eq(&self, other: &Self) -> bool {
        self.java_thread_id == other.java_thread_id
    }
}

#[derive(Debug)]
#[allow(dead_code)]
pub struct StackTrace {
    pub truncated: bool,
    pub frames: Vec<StackFrame>,
}

impl From<Accessor<'_>> for StackTrace {
    fn from(value: Accessor<'_>) -> Self {
        let truncated: bool = extract_primitive(&value, "truncated");
        let mut frames: Vec<StackFrame> = value
            .get_field("frames")
            .unwrap()
            .as_iter()
            .unwrap()
            .map(|a| a.into())
            .collect();
        frames.reverse();

        Self { truncated, frames }
    }
}

#[allow(dead_code)]
pub struct StackFrame {
    pub line_number: i32,
    pub bytecode_index: i32,
    pub method: Method,
}

impl From<Accessor<'_>> for StackFrame {
    fn from(value: Accessor<'_>) -> Self {
        let bytecode_index: i32 = extract_primitive(&value, "bytecodeIndex");
        let line_number: i32 = extract_primitive(&value, "lineNumber");
        let method: Method = value.get_field("method").map(|v| v.into()).unwrap();
        Self {
            method,
            line_number,
            bytecode_index,
        }
    }
}

impl std::fmt::Debug for StackFrame {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{:?}:{}", self.method, self.line_number)
    }
}

#[allow(dead_code)]
#[derive(Hash, PartialEq, Eq, Clone)]
pub struct Method {
    name: String,
    class: Class,
}

impl std::fmt::Debug for Method {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}:{}", self.class.pretty_name(), self.name)
    }
}

impl From<Accessor<'_>> for Method {
    fn from(value: Accessor<'_>) -> Self {
        let name = extract_symbol(&value, "name");
        let class: Class = value.get_field("type").map(|v| v.into()).unwrap();
        return Self { name, class };
    }
}

#[derive(Debug, Hash, PartialEq, Eq, Clone)]
#[allow(dead_code)]
pub struct Class {
    name: String,
}

impl Class {
    pub fn pretty_name(&self) -> String {
        self.name.replace('/', ".")
    }
}

impl std::fmt::Display for Class {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.name)
    }
}

impl From<Accessor<'_>> for Class {
    fn from(value: Accessor<'_>) -> Self {
        Self {
            name: extract_symbol(&value, "name"),
        }
    }
}

fn extract_symbol(value: &Accessor, name: &str) -> String {
    let str: &str = extract_primitive(&value.get_field(name).unwrap(), "string");
    str.to_string()
}

fn extract_nullable_str(value: &Accessor, name: &str) -> Option<String> {
    value.get_field(name)
        .ok_or_else(|| anyhow!("Unable to find field {name}"))
        .map(|v| {
            if let ValueDescriptor::Primitive(Primitive::NullString) = v.value {
                None
            } else {
                let result = <&str>::try_from(v.value)
                    .map_err(|_| anyhow!("Unable to convert {name} to string"))
                    .unwrap();
                Some(result.to_string())
            }
        })
        .unwrap()
}

fn extract_primitive<'a, T>(value: &Accessor<'a>, name: &str) -> T
where
    T: TryFrom<&'a ValueDescriptor> + Default,
{
    value
        .get_field(name)
        .and_then(|v| <T>::try_from(v.value).ok())
        .unwrap()
}
