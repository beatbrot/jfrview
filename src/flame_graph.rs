use indexmap::IndexMap;
use jfrs::reader::JfrReader;

use crate::data::{ExecutionSample, Method};
use std::fs::File;

#[derive(Default, Debug)]
pub struct FlameGraph {
    depth: usize,
    pub frames: IndexMap<Method, Frame>,
}

impl From<File> for FlameGraph {
    fn from(value: File) -> Self {
        let mut reader = JfrReader::new(value);

        let mut fg = FlameGraph::default();

        for (mut r, c) in reader.chunks().flatten() {
            r.events(&c)
                .flatten()
                .filter(|e| e.class.name() == "jdk.ExecutionSample")
                .map(|e| ExecutionSample::from(e))
                .filter(|e| !e.stack_trace.truncated)
                .for_each(|s| fg.add_sample(s));
        }
        return fg;
    }
}

#[derive(Debug, Clone)]
pub struct Frame {
    pub method: Method,
    pub ticks: usize,
    pub children: IndexMap<Method, Frame>,
}

impl Frame {
    fn new(method: Method) -> Self {
        Self {
            method,
            ticks: Default::default(),
            children: Default::default(),
        }
    }
}

impl FlameGraph {
    pub fn add_sample(&mut self, sample: ExecutionSample) {
        let mut cframe = &mut self.frames;
        let mut depth: usize = 0;
        for frame in sample.stack_trace.frames {
            let entry = cframe
                .entry(frame.method.clone())
                .or_insert_with(|| Frame::new(frame.method.clone()));
            depth += 1;
            self.depth = std::cmp::max(self.depth, depth);
            entry.ticks += 1;
            cframe = &mut entry.children;
        }
    }
}

impl std::fmt::Display for FlameGraph {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        fn printer(
            f: &mut std::fmt::Formatter<'_>,
            indent: usize,
            mt: &Method,
            frame: &Frame,
        ) -> std::fmt::Result {
            writeln!(f, "{}{:?}: {}", "| ".repeat(indent), mt, frame.ticks)?;
            for (k, v) in &frame.children {
                printer(f, indent + 1, k, v)?;
            }
            Ok(())
        }
        for (k, v) in &self.frames {
            printer(f, 0, k, v)?;
        }
        Ok(())
    }
}
