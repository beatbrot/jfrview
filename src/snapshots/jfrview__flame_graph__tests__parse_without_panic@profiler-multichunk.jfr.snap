---
source: src/flame_graph.rs
description: profiler-multichunk.jfr
input_file: test-data/profiler-multichunk.jfr
---
:/usr/lib/aarch64-linux-gnu/libc.so.6: (5706,0)
| :/usr/lib/aarch64-linux-gnu/libc.so.6: (5706,0)
| | libjvm.so:thread_native_entry: (5706,0)
| | | libjvm.so:Thread::call_run: (5706,0)
| | | | libjvm.so:ConcurrentGCThread::run: (900,0)
| | | | | libjvm.so:G1ConcurrentMarkThread::run_service: (300,0)
| | | | | | libjvm.so:Monitor::wait: (300,0)
| | | | | | | libjvm.so:Monitor::IWait: (300,0)
| | | | | | | | libjvm.so:os::PlatformEvent::park: (300,0)
| | | | | | | | | libc.so.6:pthread_cond_wait: (300,0)
| | | | | | | | | | :/usr/lib/aarch64-linux-gnu/libc.so.6: (300,0)
| | | | | libjvm.so:G1ConcurrentRefineThread::run_service: (300,0)
| | | | | | libjvm.so:Monitor::wait: (300,0)
| | | | | | | libjvm.so:Monitor::IWait: (300,0)
| | | | | | | | libjvm.so:os::PlatformEvent::park: (300,0)
| | | | | | | | | libc.so.6:pthread_cond_wait: (300,0)
| | | | | | | | | | :/usr/lib/aarch64-linux-gnu/libc.so.6: (300,0)
| | | | | libjvm.so:G1YoungRemSetSamplingThread::run_service: (300,0)
| | | | | | libjvm.so:Monitor::wait: (300,0)
| | | | | | | libjvm.so:Monitor::IWait: (300,0)
| | | | | | | | libjvm.so:os::PlatformEvent::park: (300,0)
| | | | | | | | | libc.so.6:pthread_cond_timedwait: (300,0)
| | | | | | | | | | :/usr/lib/aarch64-linux-gnu/libc.so.6: (300,0)
| | | | libjvm.so:JavaThread::thread_main_inner: (2405,0)
| | | | | libjvm.so:ServiceThread::service_thread_entry: (301,0)
| | | | | | libjvm.so:Monitor::wait: (301,0)
| | | | | | | libjvm.so:Monitor::IWait: (301,0)
| | | | | | | | libjvm.so:os::PlatformEvent::park: (301,0)
| | | | | | | | | libc.so.6:pthread_cond_wait: (301,0)
| | | | | | | | | | :/usr/lib/aarch64-linux-gnu/libc.so.6: (301,0)
| | | | | libjvm.so:CompileBroker::compiler_thread_loop: (602,0)
| | | | | | libjvm.so:CompileQueue::get: (602,0)
| | | | | | | libjvm.so:Monitor::wait: (602,0)
| | | | | | | | libjvm.so:Monitor::IWait: (602,0)
| | | | | | | | | libjvm.so:os::PlatformEvent::park: (602,0)
| | | | | | | | | | libc.so.6:pthread_cond_timedwait: (602,0)
| | | | | | | | | | | :/usr/lib/aarch64-linux-gnu/libc.so.6: (602,0)
| | | | | libjvm.so:attach_listener_thread_entry: (300,0)
| | | | | | libjvm.so:AttachListener::dequeue: (300,0)
| | | | | | | libjvm.so:LinuxAttachListener::dequeue: (300,0)
| | | | | | | | libc.so.6:accept: (300,0)
| | | | | libjvm.so:signal_thread_entry: (301,0)
| | | | | | libjvm.so:os::signal_wait: (301,0)
| | | | | | | libjvm.so:PosixSemaphore::wait: (301,0)
| | | | | | | | :/usr/lib/aarch64-linux-gnu/libc.so.6: (301,0)
| | | | | | | | | :/usr/lib/aarch64-linux-gnu/libc.so.6: (301,0)
| | | | | libjvm.so:recorderthread_entry: (300,0)
| | | | | | libjvm.so:Monitor::wait: (300,0)
| | | | | | | libjvm.so:Monitor::IWait: (300,0)
| | | | | | | | libjvm.so:os::PlatformEvent::park: (300,0)
| | | | | | | | | libc.so.6:pthread_cond_wait: (300,0)
| | | | | | | | | | :/usr/lib/aarch64-linux-gnu/libc.so.6: (300,0)
| | | | | libjvm.so:JvmtiAgentThread::call_start_function: (300,0)
| | | | | | libasyncProfiler.so:Profiler::timerLoop: (300,0)
| | | | | | | libc.so.6:pthread_cond_timedwait: (300,0)
| | | | | | | | :/usr/lib/aarch64-linux-gnu/libc.so.6: (300,0)
| | | | | libjvm.so:NMethodSweeper::sweeper_loop: (301,0)
| | | | | | libjvm.so:Monitor::wait: (301,0)
| | | | | | | libjvm.so:Monitor::IWait: (301,0)
| | | | | | | | libjvm.so:os::PlatformEvent::park: (301,0)
| | | | | | | | | libc.so.6:pthread_cond_timedwait: (301,0)
| | | | | | | | | | :/usr/lib/aarch64-linux-gnu/libc.so.6: (301,0)
| | | | libjvm.so:GangWorker::loop: (1500,0)
| | | | | libjvm.so:SemaphoreGangTaskDispatcher::worker_wait_for_task: (1500,0)
| | | | | | libjvm.so:PosixSemaphore::wait: (1500,0)
| | | | | | | :/usr/lib/aarch64-linux-gnu/libc.so.6: (1500,0)
| | | | | | | | :/usr/lib/aarch64-linux-gnu/libc.so.6: (1500,0)
| | | | libjvm.so:VMThread::run: (300,0)
| | | | | libjvm.so:VMThread::loop: (300,0)
| | | | | | libjvm.so:Monitor::wait: (300,0)
| | | | | | | libjvm.so:Monitor::IWait: (300,0)
| | | | | | | | libjvm.so:os::PlatformEvent::park: (300,0)
| | | | | | | | | libc.so.6:pthread_cond_timedwait: (300,0)
| | | | | | | | | | :/usr/lib/aarch64-linux-gnu/libc.so.6: (300,0)
| | | | libjvm.so:WatcherThread::run: (301,0)
| | | | | libjvm.so:WatcherThread::sleep: (301,0)
| | | | | | libjvm.so:Monitor::wait: (301,0)
| | | | | | | libjvm.so:Monitor::IWait: (301,0)
| | | | | | | | libjvm.so:os::PlatformEvent::park: (301,0)
| | | | | | | | | libc.so.6:pthread_cond_timedwait: (301,0)
| | | | | | | | | | :/usr/lib/aarch64-linux-gnu/libc.so.6: (301,0)
| | | | libjvm.so:JfrThreadSampler::run: (300,0)
| | | | | libjvm.so:PosixSemaphore::wait: (300,0)
| | | | | | :/usr/lib/aarch64-linux-gnu/libc.so.6: (300,0)
| | | | | | | :/usr/lib/aarch64-linux-gnu/libc.so.6: (300,0)
java.lang.Thread:run: (1379,0)
| java.util.concurrent.ThreadPoolExecutor$Worker:run: (1079,0)
| | java.util.concurrent.ThreadPoolExecutor:runWorker: (1079,0)
| | | Example$$Lambda$5.575335780:run: (599,0)
| | | | Example:lambda$main$3: (599,0)
| | | | | Example:sleep: (599,0)
| | | | | | java.lang.Thread:yield: (599,0)
| | | | | | | libc.so.6:__sched_yield: (598,0)
| | | | | | | libjvm.so:sched_yield@plt: (1,0)
| | | Example$$Lambda$4.1500056228:run: (480,0)
| | | | Example:lambda$main$2: (480,0)
| | | | | Example$Writer:println: (60,0)
| | | | | | sun.nio.ch.FileChannelImpl:force: (60,0)
| | | | | | | sun.nio.ch.FileDispatcherImpl:force: (60,0)
| | | | | | | | sun.nio.ch.FileDispatcherImpl:force0: (60,0)
| | | | | | | | | libnio.so:Java_sun_nio_ch_FileDispatcherImpl_force0: (60,0)
| | | | | | | | | | libc.so.6:fsync: (60,0)
| | | | | java.util.concurrent.LinkedBlockingQueue:poll: (420,0)
| | | | | | java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject:awaitNanos: (420,0)
| | | | | | | java.util.concurrent.locks.LockSupport:parkNanos: (420,0)
| | | | | | | | jdk.internal.misc.Unsafe:park: (420,0)
| | | | | | | | | libjvm.so:Unsafe_Park: (420,0)
| | | | | | | | | | libjvm.so:Parker::park: (420,0)
| | | | | | | | | | | libc.so.6:pthread_cond_timedwait: (420,0)
| | | | | | | | | | | | :/usr/lib/aarch64-linux-gnu/libc.so.6: (420,0)
| jdk.jfr.internal.PlatformRecorder$$Lambda$55.205465701:run: (300,0)
| | jdk.jfr.internal.PlatformRecorder:lambda$startDiskMonitor$1: (300,0)
| | | jdk.jfr.internal.PlatformRecorder:periodicTask: (300,0)
| | | | jdk.jfr.internal.PlatformRecorder:takeNap: (300,0)
| | | | | java.lang.Object:wait: (300,0)
| | | | | | libjvm.so:JVM_MonitorWait: (300,0)
| | | | | | | libjvm.so:ObjectSynchronizer::wait: (300,0)
| | | | | | | | libjvm.so:ObjectMonitor::wait: (300,0)
| | | | | | | | | libjvm.so:os::PlatformEvent::park: (300,0)
| | | | | | | | | | libc.so.6:pthread_cond_timedwait: (299,0)
| | | | | | | | | | | :/usr/lib/aarch64-linux-gnu/libc.so.6: (299,0)
| | | | | | | | | | :/usr/lib/aarch64-linux-gnu/libc.so.6: (1,0)
:unknown_Java: (1,0)
| libc.so.6:clock_gettime: (1,0)
java.lang.ref.Finalizer$FinalizerThread:run: (301,0)
| java.lang.ref.ReferenceQueue:remove: (301,0)
| | java.lang.ref.ReferenceQueue:remove: (301,0)
| | | java.lang.Object:wait: (301,0)
| | | | libjvm.so:JVM_MonitorWait: (301,0)
| | | | | libjvm.so:ObjectSynchronizer::wait: (301,0)
| | | | | | libjvm.so:ObjectMonitor::wait: (301,0)
| | | | | | | libjvm.so:os::PlatformEvent::park: (301,0)
| | | | | | | | libc.so.6:pthread_cond_wait: (301,0)
| | | | | | | | | :/usr/lib/aarch64-linux-gnu/libc.so.6: (301,0)
Example:main: (300,0)
| java.util.concurrent.ThreadPoolExecutor:awaitTermination: (300,0)
| | java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject:awaitNanos: (300,0)
| | | java.util.concurrent.locks.LockSupport:parkNanos: (300,0)
| | | | jdk.internal.misc.Unsafe:park: (300,0)
| | | | | libjvm.so:Unsafe_Park: (300,0)
| | | | | | libjvm.so:Parker::park: (300,0)
| | | | | | | libc.so.6:pthread_cond_timedwait: (300,0)
| | | | | | | | :/usr/lib/aarch64-linux-gnu/libc.so.6: (300,0)
jdk.internal.misc.InnocuousThread:run: (300,0)
| java.lang.Thread:run: (300,0)
| | jdk.internal.ref.CleanerImpl:run: (300,0)
| | | java.lang.ref.ReferenceQueue:remove: (300,0)
| | | | java.lang.Object:wait: (300,0)
| | | | | libjvm.so:JVM_MonitorWait: (300,0)
| | | | | | libjvm.so:ObjectSynchronizer::wait: (300,0)
| | | | | | | libjvm.so:ObjectMonitor::wait: (300,0)
| | | | | | | | libjvm.so:os::PlatformEvent::park: (300,0)
| | | | | | | | | libc.so.6:pthread_cond_timedwait: (300,0)
| | | | | | | | | | :/usr/lib/aarch64-linux-gnu/libc.so.6: (300,0)
:unknown: (300,0)
| java:_start: (300,0)
| | libc.so.6:__libc_start_main: (300,0)
| | | :/usr/lib/aarch64-linux-gnu/libc.so.6: (300,0)
| | | | java:main: (300,0)
| | | | | libjli.so:JLI_Launch: (300,0)
| | | | | | libjli.so:ContinueInNewThread: (300,0)
| | | | | | | libjli.so:CallJavaMainInNewThread: (300,0)
| | | | | | | | :/usr/lib/aarch64-linux-gnu/libc.so.6: (300,0)
| | | | | | | | | :/usr/lib/aarch64-linux-gnu/libc.so.6: (300,0)
java.lang.ref.Reference$ReferenceHandler:run: (301,0)
| java.lang.ref.Reference:processPendingReferences: (301,0)
| | java.lang.ref.Reference:waitForReferencePendingList: (301,0)
| | | libjvm.so:JVM_WaitForReferencePendingList: (301,0)
| | | | libjvm.so:Monitor::wait: (301,0)
| | | | | libjvm.so:Monitor::IWait: (301,0)
| | | | | | libjvm.so:os::PlatformEvent::park: (301,0)
| | | | | | | libc.so.6:pthread_cond_wait: (301,0)
| | | | | | | | :/usr/lib/aarch64-linux-gnu/libc.so.6: (301,0)
java.util.TimerThread:run: (300,0)
| java.util.TimerThread:mainLoop: (300,0)
| | java.lang.Object:wait: (300,0)
| | | java.lang.Object:wait: (300,0)
| | | | libjvm.so:JVM_MonitorWait: (300,0)
| | | | | libjvm.so:ObjectSynchronizer::wait: (300,0)
| | | | | | libjvm.so:ObjectMonitor::wait: (300,0)
| | | | | | | libjvm.so:os::PlatformEvent::park: (300,0)
| | | | | | | | libc.so.6:pthread_cond_wait: (300,0)
| | | | | | | | | :/usr/lib/aarch64-linux-gnu/libc.so.6: (300,0)
